synapse/third_party_rules/access_rules.py

```diff
@@ -0,0 +1,375 @@
+        default rule to the initial state.
+        """
+        is_direct = config.get("is_direct")
+        rules_in_initial_state = False
+
+        # If there's a rules event in the initial state, check if it complies with the
+        # spec for im.vector.room.access_rules and fix it if not.
+        for event in config.get("initial_state", []):
+            if event["type"] == ACCESS_RULES_TYPE:
+                rules_in_initial_state = True
+
+                rule = event["content"].get("rule")
+
+                # Make sure the event has a valid content.
+                if rule is None:
+                    event["content"] = {
+                        "rule": self._on_create_room_default_rule(is_direct)
+                    }
+
+                # Make sure the rule name is valid.
+                if not self._is_rule_name_valid(rule):
+                    event["content"]["rule"] = self._on_create_room_default_rule(
+                        is_direct,
+                    )
+
+                # Make sure the rule is "direct" if the room is a direct chat.
+                if is_direct and rule != ACCESS_RULE_DIRECT:
+                    event["content"]["rule"] = ACCESS_RULE_DIRECT
+
+                # Make sure the rule is not "direct" if the room isn't a direct chat.
+                if rule == ACCESS_RULE_DIRECT and not is_direct:
+                    event["content"]["rule"] = ACCESS_RULE_RESTRICTED
+
+        # If there's no rules event in the initial state, create one with the default
+        # setting.
+        if not rules_in_initial_state:
+            config["initial_state"].append({
+                "type": ACCESS_RULES_TYPE,
+                "state_key": "",
+                "content": {
+                    "rule": self._on_create_room_default_rule(is_direct),
+                }
+            })
+
+    @staticmethod
+    def _on_create_room_default_rule(is_direct):
+        """Returns the default rule to set.
+
+        Args:
+            is_direct (bool): Is the room created with "is_direct" set to True.
+
+        Returns:
+            str, the name of the rule tu use as the default.
+        """
+        if is_direct:
+            return ACCESS_RULE_DIRECT
+        else:
+            return ACCESS_RULE_RESTRICTED
+
+    @defer.inlineCallbacks
+    def check_threepid_can_be_invited(self, medium, address, state_events):
+        """Implements synapse.events.ThirdPartyEventRules.check_threepid_can_be_invited
+
+        Check if a threepid can be invited to the room via a 3PID invite given the current
+        rules and the threepid's address, by retrieving the HS it's mapped to from the
+        configured identity server, and checking if we can invite users from it.
+        """
+        rule = self._get_rule_from_state(state_events)
+
+        if medium != "email":
+            defer.returnValue(False)
+
+        if rule != ACCESS_RULE_RESTRICTED:
+            # Only "restricted" requires filtering 3PID invites.
```

Comment:
Can you add a note that direct chat membership filtering will happen later so doesn't need to be done here?